### PR TITLE
Support Cython 3.3

### PR DESCRIPTION
**Pull request by antonio-rojas**
See individual commit messages for details

### CODE REVIEW

**Comment by kiwifb** [comment]
It built. Let's wait for the bots to catch up.

**Comment by kiwifb** [comment]
Build but crash immediately on start up here. Fairly big backtrace. Top if that's helpful
```
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
__syscall_cancel_arch ()
    at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56

Stack backtrace
---------------
    at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56
No locals.
#1  0x00007f94ab0a4088 in __internal_syscall_cancel (a1=<optimized out>, 
    a2=a2@entry=0, a3=a3@entry=0, a4=a4@entry=0, a5=a5@entry=0, a6=a6@entry=0,
    nr=61) at cancellation.c:49
        result = <optimized out>
        pd = <optimized out>
        ch = <optimized out>
#2  0x00007f94ab0a40d1 in __syscall_cancel (a1=<optimized out>, a2=a2@entry=0, 
    a3=a3@entry=0, a4=a4@entry=0, a5=a5@entry=0, a6=a6@entry=0, nr=61)
    at cancellation.c:75
        r = <optimized out>
#3  0x00007f94ab11005b in __GI___wait4 (pid=<optimized out>, 
    stat_loc=stat_loc@entry=0x0, options=options@entry=0,
    usage=usage@entry=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
No locals.
#4  0x00007f94ab11009b in __GI___waitpid (pid=<optimized out>, 
    stat_loc=stat_loc@entry=0x0, options=options@entry=0) at waitpid.c:38
No locals.
#5  0x00007f94aa736bf3 in print_enhanced_backtrace ()
    at ../cysignals-1.12.6/src/cysignals/implementation.c:780
        parent_pid = 24063
        pid = <optimized out>
#6  sigdie (sig=sig@entry=11, 
    s=s@entry=0x7f94aa73e578 "Unhandled SIGSEGV: A segmentation fault occurred.") at ../cysignals-1.12.6/src/cysignals/implementation.c:806
No locals.
#7  0x00007f94aa736f5a in sigdie_for_sig (sig=11, inside=0)
    at ../cysignals-1.12.6/src/cysignals/implementation.c:262
No locals.
#8  cysigs_signal_handler (sig=11)
    at ../cysignals-1.12.6/src/cysignals/implementation.c:464
        inside = <optimized out>
#9  <signal handler called>
No locals.
#10 0x00007f94aa3b84d9 in ?? ()
   from /usr/lib/python3.14/site-packages/sage/structure/sage_object.cpython-314-x86_64-linux-gnu.so
No symbol table info available.
#11 0x00007f94aa45ab65 in ?? ()
   from /usr/lib/python3.14/site-packages/sage/structure/element.cpython-314-x86_64-linux-gnu.so
No symbol table info available.
#12 0x00007f94a8945662 in ?? ()
   from /usr/lib/python3.14/site-packages/sage/rings/rational.cpython-314-x86_64-linux-gnu.so
No symbol table info available.
```

**Comment by antonio-rojas** [comment]
> Build but crash immediately on start up here. Fairly big backtrace. Top if that's helpful

No idea, works fine here. Maybe debug symbols for sagelib would give a hint.

I've updated cython in conda, let's see what CI says

**Comment by kiwifb** [comment]
I suspect all packages using cython have to be rebuilt but there maybe more to it. Maybe some of my dependencies are not currently correct.

**Comment by antonio-rojas** [comment]
I didn't rebuild anything besides sagelib

**Comment by kiwifb** [comment]
Curious. It is too late locally to do anymore today.

**Comment by antonio-rojas** [comment]
Conda CI seems happy. Would be good to find out what's failing on your system.

**Comment by kiwifb** [comment]
It is very strange, if I compile with cython 3.2.9 sage starts. It is really cython 3.3.0 specific. Make me wonder if there is a problem with cython 3.3.0 on Gentoo.

**Comment by cxzhong** [comment]
maybe we need to build cysignals with cython 3.3+

**Comment by antonio-rojas** [comment]
> maybe we need to build cysignals with cython 3.3+

Could be tried, but I doubt it would help. Neither me or the conda CI runners use cysignals built with cython 3.3. What could be useful is a backtrace with debug symbols for /usr/lib/python3.14/site-packages/sage/structure/sage_object.cpython-314-x86_64-linux-gnu.so, so the exact source of the crash can be traced.

File: src/sage/cpython/cython_metaclass.h [hunk 1]
```diff
@@ -45,6 +49,12 @@ static CYTHON_INLINE int Sage_PyType_Ready(PyTypeObject* t)
     if (r < 0)
         return r;
 
+#if defined(CYTHON_HEX_VERSION) && CYTHON_HEX_VERSION >= 0x03030000
+    // Ensure Cython function types are initialized before calling metaclass methods.
+    if (__Pyx_InitAfterSharedUtility() < 0)
```
Comment (cxzhong) [inline src/sage/cpython/cython_metaclass.h:54]: `__Pyx_InitAfterSharedUtility()` is not idempotent in Cython 3.3. `Sage_PyType_Ready()` can run once per extension type, and the Cython-generated module initialization calls this helper again after type initialization. The helper CyFunction/FusedFunction initialization stores a newly owned reference in module state without dropping the previous one, so these repeated calls leak type references. It also couples Sage to a private generated helper whose contents may change in later Cython versions. Could this use an idempotent or once-only upstream hook, with a regression test covering a module containing multiple extension types, instead of invoking the whole helper for every type?

File: src/sage/data_structures/list_of_pairs.pyx [hunk 2]
```diff
@@ -67,10 +67,6 @@ cdef class ListOfPairs:
             Traceback (most recent call last):
             ...
             IndexError
```
Comment (cxzhong) [inline src/sage/data_structures/list_of_pairs.pyx:69]: Please keep the negative-index regression test and update its expected exception from `OverflowError` to `IndexError`, rather than deleting it. Without `l[-1]`, this no longer checks that conversion to `size_t` rejects a negative index instead of underflowing or reaching an unintended element. It would also be useful to cover the corresponding negative-index `__setitem__` path.

**Comment by kiwifb** [comment]
> maybe we need to build cysignals with cython 3.3+

I actually have done so and it still crashes. In fact I had cysignals built with cython 3.3.0 and sagemath with cython 3.2.9 and that worked.
I'll see what I can do about debugging symbols in the morning.

File: src/sage/data_structures/list_of_pairs.pyx (hunk 2 above)
Comment (antonio-rojas) [reply src/sage/data_structures/list_of_pairs.pyx:69]: We can't have different expected results for different Cython versions. Once cython 3.3 is required we can add a test that verifies that negative indices work as expected.

**Comment by kiwifb** [comment]
So a better backtrace
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=11, no_tid=no_tid@entry=0) at pthread_kill.c:44

⚠ warning: 44  pthread_kill.c: No such file or directory
(gdb) bt full
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=11, no_tid=no_tid@entry=0) at pthread_kill.c:44
        tid = <optimized out>
        ret = 0
        pd = <optimized out>
        old_mask = {__val = {1}}
        ret = <optimized out>
#1  0x00007f410a4a9413 in __pthread_kill_internal (threadid=<optimized out>, signo=11) at pthread_kill.c:89
No locals.
#2  0x00007f410a451616 in __GI_raise (sig=sig@entry=11) at ../sysdeps/posix/raise.c:26
        ret = <optimized out>
#3  0x00007f4109b4eb69 in sigdie (sig=sig@entry=11, s=s@entry=0x7f4109b56578 "Unhandled SIGSEGV: A segmentation fault occurred.")
    at ../cysignals-1.12.6/src/cysignals/implementation.c:821
No locals.
#4  0x00007f4109b4ef5a in sigdie_for_sig (sig=11, inside=0) at ../cysignals-1.12.6/src/cysignals/implementation.c:262
No locals.
#5  cysigs_signal_handler (sig=11) at ../cysignals-1.12.6/src/cysignals/implementation.c:464
        inside = <optimized out>
#6  <signal handler called>
No locals.
#7  _Py_TYPE (ob=0x0) at /usr/include/python3.14/object.h:277
No locals.
#8  PyTuple_GET_SIZE (op=0x0) at /usr/include/python3.14/cpython/tupleobject.h:24
        tuple = <optimized out>
        __PRETTY_FUNCTION__ = "PyTuple_GET_SIZE"
#9  __Pyx_CallTpnewAsVectorcall (f=<optimized out>, o=0x7f41037fb920 <__pyx_type_4sage_5rings_8rational_Rational>, a=0x0, k=0x0)
    at src/sage/structure/sage_object.cpython-314-x86_64-linux-gnu.so.p/src/sage/structure/sage_object.pyx.c:20694
        k_size = 0
        a_size = <optimized out>
        stack_args = {'infinity', <unknown at remote 0x7ffc03208cf0>, 0x0, <unknown at remote 0x7f410a6ef52a>, 
          {'__name__': 'sage.rings.padics.common_conversion', '__doc__': 'File: sage/rings/padics/common_conversion.pyx (starting at line 1)\n\nTools for creation of `p`-adic elements\n\nThe functions in this file are used in creating new `p`-adic elements.\n\nWhen creating a `p`-adic element, the user can specify that the absolute\nprecision be bounded and/or that the relative precision be bounded.\nMoreover, different `p`-adic parents impose their own bounds on the\nrelative or absolute precision of their elements.  The precision\ndetermines to what power of `p` the defining data will be reduced, but\nthe valuation of the resulting element needs to be determined before\nthe element is created.  Moreover, some defining data can impose their\nown precision bounds on the result.\n\nAUTHORS:\n\n- David Roe (2012-03-01)\n', '__package__': 'sage.rings.padics', '__loader__': <ExtensionFileLoader(name='sage.rings.padics.common_conversion', path='/usr/lib/python3.14/site-packages/sage/rings/padics/common_conversion.cpython-31...(truncated)}
        args = <optimized out>
        maxnargs = <optimized out>
        total_size = <optimized out>
        result = <optimized out>
        kwnames = <optimized out>
        unpack_dict_result = <optimized out>
        i = <optimized out>
        cleanup = <optimized out>
        __PRETTY_FUNCTION__ = "__Pyx_CallTpnewAsVectorcall"
#10 __pyx_tp_new_4sage_9structure_11sage_object_SageObject (t=0x7f41037fb920 <__pyx_type_4sage_5rings_8rational_Rational>, a=0x0, k=0x0)
    at src/sage/structure/sage_object.cpython-314-x86_64-linux-gnu.so.p/src/sage/structure/sage_object.pyx.c:15574
No locals.
#11 0x00007f41098a4b65 in __pyx_tp_new_4sage_9structure_7element_Element (t=<optimized out>, a=<optimized out>, k=<optimized out>)
    at src/sage/structure/element.cpython-314-x86_64-linux-gnu.so.p/src/sage/structure/element.pyx.c:40690
        o = <optimized out>
        base_tp_new = <optimized out>
#12 __pyx_tp_new_4sage_9structure_7element_ModuleElement (t=<optimized out>, a=<optimized out>, k=<optimized out>)
--Type <RET> for more, q to quit, c to continue without paging--
    at src/sage/structure/element.cpython-314-x86_64-linux-gnu.so.p/src/sage/structure/element.pyx.c:41138
        o = <optimized out>
#13 __pyx_tp_new_4sage_9structure_7element_RingElement (t=<optimized out>, a=<optimized out>, k=<optimized out>)
    at src/sage/structure/element.cpython-314-x86_64-linux-gnu.so.p/src/sage/structure/element.pyx.c:41257
        o = <optimized out>
#14 __pyx_tp_new_4sage_9structure_7element_CommutativeRingElement (t=<optimized out>, a=<optimized out>, k=<optimized out>)
    at src/sage/structure/element.cpython-314-x86_64-linux-gnu.so.p/src/sage/structure/element.pyx.c:42058
        o = <optimized out>
#15 __pyx_tp_new_4sage_9structure_7element_FieldElement (t=<optimized out>, a=<optimized out>, k=<optimized out>)
    at src/sage/structure/element.cpython-314-x86_64-linux-gnu.so.p/src/sage/structure/element.pyx.c:42643
        o = <optimized out>
#16 0x00007f41037a0662 in __pyx_tp_new_4sage_5rings_8rational_Rational (t=<optimized out>, a=<optimized out>, k=<optimized out>)
    at src/sage/rings/rational.cpython-314-x86_64-linux-gnu.so.p/src/sage/rings/rational.pyx.cpp:46204
        o = <optimized out>
        base_tp_new = <optimized out>
#17 0x00007f410140c840 in __pyx_f_4sage_3ext_7stdsage_PY_NEW (__pyx_v_t=<optimized out>)
    at src/sage/rings/padics/common_conversion.cpython-314-x86_64-linux-gnu.so.p/src/sage/rings/padics/common_conversion.pyx.c:5712
        __pyx_r = 0x0
        __pyx_t_1 = 0x0
        __pyx_lineno = 0
        __pyx_filename = 0x0
        __pyx_clineno = 0
        __pyx_L1_error = <optimized out>
        __pyx_temp = <optimized out>
#18 0x00007f410140e6a0 in __pyx_pymod_exec_common_conversion (__pyx_pyinit_module=<optimized out>)
    at src/sage/rings/padics/common_conversion.cpython-314-x86_64-linux-gnu.so.p/src/sage/rings/padics/common_conversion.pyx.c:12584
        stringtab_initialized = 1
        __pyx_mstate = 0x7f41014206e0 <__pyx_mstate_global_static>
        __pyx_t_1 = <optimized out>
        __pyx_t_2 = 0x0
        __pyx_t_3 = 0
        __pyx_t_4 = 0x0
        __pyx_lineno = 0
        __pyx_filename = 0x0
        __pyx_clineno = 0
#19 0x00007f410a6fed06 in PyModule_ExecDef (module=<module at remote 0x7f41016c11c0>, def=<optimized out>) at Objects/moduleobject.c:498
        cur_slot = 0x7f4101420310 <__pyx_moduledef_slots+16>
        name = 0x7f4101ad10e8 "sage.rings.padics.common_conversion"
        ret = <optimized out>
#20 0x00007f410a8245dd in exec_builtin_or_dynamic (mod=<optimized out>) at Python/import.c:891
        def = 0x7ffc032091d0
        state = <optimized out>
#21 0x00007f410a8245f0 in _imp_exec_dynamic_impl (module=<optimized out>, mod=<optimized out>) at Python/import.c:4840
No locals.
#22 _imp_exec_dynamic (module=<optimized out>, mod=<optimized out>) at Python/clinic/import.c.h:517
        return_value = 0x0
        _return_value = <optimized out>
#23 0x00007f410a7de94a in _PyEval_EvalFrameDefault (tstate=0x7f410ab5d140 <_PyRuntime+315680>, frame=0x7f410ac42c50, throwflag=0) at Python/generated_cases.c.h:2656
        callargs = <optimized out>
        kwargs = <optimized out>
        kwargs = <optimized out>
        arg = <optimized out>
```
Let me know if you need more of it.

**Comment by antonio-rojas** [comment]
> So a better backtrace
> 

Thanks, that's useful. Given that I get the exact same backtrace up to frame #9, I figured the difference must be the cflags, and I was able to reproduce this by dropping `-DNDEBUG` 

Looks like a cython bug, and I fear it's going to be hard to come up with a simple test case.

File: src/sage/data_structures/list_of_pairs.pyx (hunk 2 above)
Comment (vincentmacri) [reply src/sage/data_structures/list_of_pairs.pyx:69]: > We can't have different expected results for different Cython versions. Once cython 3.3 is required we can add a test that verifies that negative indices work as expected.

What about making the expected result less strict for the time being? Something like:

```
            Traceback (most recent call last):
            ...
            ...Error
```
and a note to change it to `IndexError` once Cython 3.3 is required.

**Comment by vincentmacri** [comment]
Don't know enough to help with debugging, but in case it helps I was able to successfully build this PR on my machine and the test suite passed (except for a spurious failure).

OS: Fedora 44
Install method: Conda (with conda installed from system packages)
Python version: 3.14

**Comment by antonio-rojas** [comment]
> So a better backtrace

Does it work now?

File: src/sage/data_structures/list_of_pairs.pyx (hunk 2 above)
Comment (antonio-rojas) [reply src/sage/data_structures/list_of_pairs.pyx:69]: Not sure how useful that would be, especially considering that this is an EXAMPLES block (not TESTS).

**Comment by kiwifb** [comment]
> > So a better backtrace
> 
> Does it work now?

Yes!

**Comment by cxzhong** [comment]
maybe you can wait my https://github.com/sagemath/sage/pull/41879 merged, and refresh the conda-lock to cython 3.3.0

**Comment by antonio-rojas** [comment]
> maybe you can wait my #41879 merged, and refresh the conda-lock to cython 3.3.0

My plan is to revert 44627cac79fb4750f55fd69d2f62f15d1684aa74 after this gets a positive review. The point of the version increase was just to force a CI run with 3.3.0.

**Comment by vbraun** [comment]
Great idea, but really a blocker?

**Comment by cxzhong** [comment]
can you revert the conda-lock change due to conflict?

**Comment by antonio-rojas** [comment]
> Great idea, but really a blocker?

It would be good to get this in so that rolling distros don't have to carry a patch for the next few months.

**Comment by antonio-rojas** [comment]
> can you revert the conda-lock change due to conflict?

Done

File: src/sage/cpython/cython_metaclass.h (hunk 1 above)
Comment (cxzhong) [reply src/sage/cpython/cython_metaclass.h:54]: [P3] Make the early function-type initialization idempotent

I rechecked this against Cython 3.2.9 and 3.3.0 at b19596a4. The early initialization is necessary: the original header crashes on import with 3.3.0, while this change fixes it on both CPython 3.12 and 3.14.

The unbalanced references are also reproducible. After warming the shared function type, importing a module with one marker function and 20 empty extension types increases the CyFunction reference count by 2 with 3.2.9, but by 22 with 3.3.0 and this patch. Each type-ready call invokes the helper, whose CyFunction/FusedFunction initialization overwrites an owned reference without releasing the previous one; the generated module initializer invokes it again afterward.

I would treat this as a low-priority follow-up, not a merge blocker: these are extra references to cached shared types, not a fresh type allocation on every call. Could the initialization be made idempotent, with a regression test covering multiple extension types? A once-only guard in Sage alone would still leave the later call from the generated module initializer.
